### ISSUE TITLE
[BUG] rspamd 3.11 segfaults on startup

### ISSUE THREAD

**Issue by cryptomilk** [rspamd/rspamd#5270]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
    * Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
    * Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
    * Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
    * Checked that there is not already an experimental package or master branch

**Describe the bug**
On openSUSE 15.6, rspamd crashes on startup:

```
Thread 1 "rspamd" received signal SIGSEGV, Segmentation fault.                                                                                                                                
0x00007ffff75814a9 in __strlen_avx2 () from /lib64/libc.so.6                                                                                                                                  
(gdb) bt                                                                                                                                                                                      
#0  0x00007ffff75814a9 in __strlen_avx2 () from /lib64/libc.so.6                               
#1  0x00007ffff7d5a573 in rspamd_vprintf_common (func=func@entry=0x7ffff7d568d0 <rspamd_printf_append_char(char const*, glong, gpointer)>, apd=apd@entry=0x7fffffffbae0, fmt=0x555555579974 "s
", args=<optimized out>) at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/libutil/printf.c:907                                                                                           
#2  0x00007ffff7d5ab85 in rspamd_vsnprintf (buf=<optimized out>, max=<optimized out>, fmt=<optimized out>, args=<optimized out>) at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/libutil
/printf.c:561                                                                                  
#3  0x00007ffff7ca22be in rspamd_common_logv (rspamd_log=0x5555555f7140, rspamd_log@entry=0x0, level_flags=64, module=0x5555555f6f68 "main", id=0x5555555f6f78 "1a49ae5b1505cc", function=0x55
555557e763 <__func__.49985> "main", fmt=<optimized out>, args=0x7fffffffdbe0) at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/libserver/logger/logger.c:457
#4  0x00007ffff7ca286d in rspamd_default_logv (level_flags=<optimized out>, module=<optimized out>, id=<optimized out>, function=<optimized out>, fmt=<optimized out>, args=args@entry=0x7ffff
fffdbe0) at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/libserver/logger/logger.c:542                                                                                                  
#5  0x00007ffff7ca2905 in rspamd_default_log_function (level_flags=level_flags@entry=64, module=<optimized out>, id=<optimized out>, function=function@entry=0x55555557e763 <__func__.49985> "
main", fmt=fmt@entry=0x55555557995b "simdutf implementation: %s") at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/libserver/logger/logger.c:553                                         
#6  0x00005555555614e9 in main (argc=<optimized out>, argv=<optimized out>, env=<optimized out>) at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/rspamd.c:1555

(gdb) f 6
#6  0x00005555555614e9 in main (argc=<optimized out>, argv=<optimized out>, env=<optimized out>) at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/rspamd.c:1555
1555            msg_info_main("simdutf implementation: %s", rspamd_fast_utf8_library_impl_name());
(gdb) p rspamd_fast_utf8_library_impl_name()
$19 = 0x555556622a30 "haswell(Intel/AMD AVX2)"

(gdb) f 1
#1  0x00007ffff7d5a573 in rspamd_vprintf_common (func=func@entry=0x7ffff7d568d0 <rspamd_printf_append_char(char const*, glong, gpointer)>, apd=apd@entry=0x7fffffffbae0, fmt=0x555555579974 "s", args=<optimized out>) at /usr/src/debug/rspamd-3.11.0-lp156.99.2.x86_64/src/libutil/printf.c:907
907                                                     slen = strlen(p);
(gdb) p p
$21 = 0x56622a30 <error: Cannot access memory at address 0x56622a30>
```

This is really strange. It looks like the something got removed from the memory address:
```
frame 1:     0x56622a30
frame 6: 0x555556622a30
```

I will try to build the RPM with Address Sanitizer next. The problem is that it build snowball with asan and snowball then fails to run with asan :-(

### Steps to Reproduce

1. Install openSUSE 15.6
2. zypper ar http://download.opensuse.org/repositories/server:/mail/15.6/server:mail.repo
3. zypper install rpsamd
4. systemctl start rspamd.service

**Expected behavior**
rspamd should start without segfaulting

### Versions
rspamd-3.11.0-lp156.99.2.x86_64

**Comment by cryptomilk** [comment]
The package got built with gcc-9. I've updated to gcc-14 and I get the following now:

```
[   46s] /home/abuild/rpmbuild/BUILD/rspamd-3.11.0/src/rspamd.c:1555:53: error: implicit declaration of function 'rspamd_fast_utf8_library_impl_name' [-Wimplicit-function-declaration]
[   46s]  1555 |         msg_info_main("simdutf implementation: %s", rspamd_fast_utf8_library_impl_name());
[   46s]       |                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by cryptomilk** [comment]
No, the crash is for sure because of the implicit declaration.

**Comment by cryptomilk** [comment]
Fixed with https://github.com/rspamd/rspamd/commit/ccb45df90df60fae36b9438cfb2b0088e590306b

**Comment by vstakhov** [comment]
In fact, I don’t understand why someone wants to build or run Rspamd without hyperscan or vectorscan. This bug doesn’t appear when hyperscan is enabled, and from my understanding it must be enabled all the time except some marginal platforms.Sent from my iPhoneOn 24 Dec 2024, at 17:21, Andreas Schneider ***@***.***> wrote:﻿
No, the crash is for sure because of the implicit declaration.

**Comment by darix** [comment]
for the simple reason that some older distributions didnt have hyperscan. we applied the fix from ccb45df now and also could enable hyperscan on top of it.
